File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-13.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]

        if nums[0] < nums[-1]:
            return nums[0]
        
        left, right = 0, len(nums) - 1

        while left <= right:
            mid = (left + right + 1) // 2

            if nums[mid - 1] > nums[mid]:
                return nums[mid]
            if nums[left] > nums[mid]:
                right = mid
            else:
                left = mid

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-13.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        
        Map<Character, Integer> charIndices = new HashMap();
        int left = 0;
        int maxLen = 0;

        for (int right = 0; right < s.length(); right++) {
            Character c = s.charAt(right);

            // when to shrink
            if (charIndices.containsKey(c) && charIndices.get(c) >= left) {
                left = charIndices.get(c) + 1;
            }

            maxLen = Math.max(maxLen, right - left + 1);
            charIndices.put(c, right);
        }

        return maxLen;
    }
}
